File: backend/app/schemas/voci.py
"""Contratti di `/voci`: la libreria personale e le sue Voci.

Nessun campo `id`/`utente_id` in ingresso in nessuno schema (AGENTS.md):
l'identità arriva sempre da `get_current_user`, mai dal body.
"""

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.letture import EsitoLettura
from app.schemas.recensioni import RecensioneResponse, Visibilita

StatoVoce = Literal["da_leggere", "in_lettura", "in_pausa", "abbandonato", "letto"]


class AggiungiVoceRequest(BaseModel):
    """Corpo di POST /voci. `libro_id` deve già esistere in `public.libro`
    (in questa issue trattato come dato seminato fuori banda — la ricerca
    sui cataloghi esterni che lo popola è l'issue #4)."""

    libro_id: UUID


class AutoreEssenziale(BaseModel):
    """Identità stabile dell'Autore (ADR 0005): il Libro punta a questa,
    mai a un nome scritto a mano. Dato condiviso, sola lettura qui."""

    id: UUID
    nome_canonico: str


class GenereEssenziale(BaseModel):
    """Identità stabile del Genere (ADR 0005) più l'etichetta nella lingua
    dell'interfaccia — oggi sempre italiano, l'interfaccia bilingue resta
    debito noto (AGENTS.md). Dato condiviso, nessuna correzione da app
    (regola 21/22 del PRD)."""

    id: str
    etichetta: str


class LibroDaScaffale(BaseModel):
    """Il sottoinsieme del Libro che serve allo SCAFFALE:
    dato condiviso, sola lettura in questa issue. `autori` arriva già
    ordinato per `ordine` (PRD regola 18, peso ripartito tra più autori)
    — vedi `app/repositories/voce_repository.py::_appiattisci_autori`."""

    id: UUID
    titolo_canonico: str
    anno_prima_pubblicazione: int | None
    anno_dedotto: bool
    """Vero quando il valore viene dal modello e non dal catalogo/Wikidata
    (issue #20 punti 1+2). L'etichetta "dedotto" prevista da design §9 è
    stata costruita e poi tolta dall'interfaccia (emendamento 22 agosto
    2026): il campo resta esposto, non più mostrato in scheda."""
    lingua_originale: str | None
    lingua_dedotta: bool
    """Come `anno_dedotto`, per la lingua originale."""
    generi: list[GenereEssenziale]
    """Fino a tre (PRD), mai correggibili da app (regola 21/22): "nessun
    affordance di modifica" è il messaggio, non solo l'assenza di un
    comando (design §9)."""
    copertina_miniatura_url: str | None
    """URL firmato, non il percorso interno: il bucket è privato (PRD
    regola 6) e il percorso da solo non apre nulla. La firma dura sette
    giorni ed è stabile tra le richieste, così il browser può davvero
    metterla in cache (app/core/storage.py)."""
    copertina_grande_url: str | None
    copertina_colore_dominante: str | None
    """Governa ombra e fondo del volume sullo scaffale (design §7).
    Assente finché la copertina non è stata recuperata."""
    copertina_colore_dominante_scuro: str | None
    """Variante desaturata di `copertina_colore_dominante` per la stanza
    scura (design §3). Assente esattamente quando lo è la prima."""
    copertina_stato: str
    """`in_attesa` mentre il lavoro in secondo piano sta lavorando: è ciò
    che permette allo scaffale di sapere che vale la pena ricontrollare,
    invece di aspettare per sempre un'immagine che forse non arriverà
    (PRD, "lavori in secondo piano con uno stato osservabile")."""
    autori: list[AutoreEssenziale]


class LibroEssenziale(LibroDaScaffale):
    """Il Libro come lo vuole la SCHEDA (`GET /voci/{id}`): tutto quanto
    sopra, più la descrizione dell'opera.

    La descrizione sta qui e non in `LibroDaScaffale` perché lo scaffale
    non la disegna in nessun punto — `components/libreria/` non la
    nomina — mentre `GET /voci` la spediva per ogni volume della
    libreria. Su una libreria di qualche centinaio di titoli era la voce
    più pesante di tutta la risposta, scaricata a ogni apertura della
    home e mai letta. Non è una restrizione dell\'API: è la stessa
    distinzione che l\'interfaccia fa già da sé fra la fila dei dorsi e
    la carta aperta di un libro."""

    descrizione: str | None
    """Solo nella lingua dell'interfaccia, mai un ripiego su un'altra
    (design §9): assente se quella lingua non ha una descrizione, anche
    se un'altra ce l'ha."""
    descrizione_riformulata: bool
    """Vero quando il testo è stato riformulato dal modello (espanso se
    troppo corto, accorciato se troppo lungo) a partire dalla descrizione
    sorgente (design §24, emendamento 21 agosto 2026). L'etichetta di
    trasparenza in scheda è stata costruita e poi tolta dall'interfaccia
    (emendamento 22 agosto 2026): il campo resta esposto, non più
    distinto in scheda dalla citazione letterale della fonte."""


class VoceResponse(BaseModel):
    id: UUID
    # Proprietario della Voce: da quando GET /voci/{id} è raggiungibile
    # anche da un collegato attivo (issue #3), il frontend usa questo
    # campo per distinguere "è mia" da "è di un collegato" nella stessa
    # pagina condivisa e nascondere i controlli di scrittura di
    # conseguenza (design-frontend.md §15).
    utente_id: UUID
    libro_id: UUID
    stato: StatoVoce
    pagine_adottate: int | None
    # Mezze stelle ammesse (design doc §9): 1, 1.5, 2, ..., 5.
    voto: float | None
    nota_intenzione: str | None
    creato_at: datetime
    aggiornato_at: datetime
    # Pagina dell'ultimo avanzamento della Lettura aperta, se c'è: solo
    # GET /voci la valorizza davvero (design-frontend.md §7, filo di
    # avanzamento sulla fascia "in corso"); altrove resta None per
    # costruzione della query, non per assenza di dato.
    pagina_corrente: int | None = None
    # Conteggi, non contenuto: un conteggio non è un elenco o un'anteprima
    # ai fini della regola 10 del PRD, quindi nessun gating spoiler qui —
    # a differenza del testo vero e proprio di recensione/insight, esposto
    # solo da GET /voci/{id} con le sue regole di visibilità (issue #5).
    # Per "Nella tua libreria".
    ha_recensione: bool = False
    numero_insight: int = 0


class VoceConLibroResponse(VoceResponse):
    """Per lo scaffale e la scheda: la Voce con il Libro incorporato,
    così il frontend non deve fare una seconda chiamata per ogni dorso."""

    libro: LibroDaScaffale


class AvanzamentoEssenziale(BaseModel):
    """Il sottoinsieme dell'Avanzamento incorporato nella scheda del
    libro (GET /voci/{id}): niente `lettura_id`, già implicito
    nell'annidamento sotto la sua Lettura."""

    id: UUID
    pagina: int
    data: date
    generato_automaticamente: bool


class InsightEssenziale(BaseModel):
    """Annidato in `GET /voci/{id}` (`letture[].insight` e
    `insight_senza_lettura`, issue #5). `testo` è `None` se e solo se
    `spoiler` è vero **e chi guarda non è il proprietario** — un collegato
    in visione reciproca lo vede tagliato, il proprietario vede sempre il
    testo pieno (design-frontend.md §11, rivisto nell'issue #6: la regola
    10 protegge da uno spoiler altrui, non da un proprio testo). Per il
    collegato, il testo pieno si ottiene solo con `GET /insight/{id}/testo`,
    dietro un gesto esplicito."""

    id: UUID
    testo: str | None
    spoiler: bool
    visibilita: Visibilita
    data: date
    creato_at: datetime


class LetturaConAvanzamenti(BaseModel):
    """Una Lettura con i propri avanzamenti, per lo storico delle
    letture nella scheda del libro (design-frontend.md §9, "sotto le due
    pagine")."""

    id: UUID
    data_inizio: date
    data_fine: date | None
    esito: EsitoLettura | None
    avanzamenti: list[AvanzamentoEssenziale]
    # Raggruppati per Lettura (design-frontend.md §10, "come impone il
    # PRD"), gating spoiler già applicato — issue #5.
    insight: list[InsightEssenziale]


class VoceDettaglioResponse(VoceResponse):
    """GET /voci/{id}: la Voce con il Libro e l'intero storico delle
    Letture (ciascuna con i propri avanzamenti) — quanto serve alla
    scheda del libro senza ulteriori chiamate."""

    libro: LibroEssenziale
    letture: list[LetturaConAvanzamenti]
    # `None` se non scritta, o se scritta ma privata e chi guarda non è il
    # proprietario (RLS, non un campo booleano applicativo) — issue #5.
    recensione: RecensioneResponse | None
    # Insight non legati a nessuna Lettura: scritti prima di iniziare il
    # libro, o orfani di una Lettura poi cancellata (PRD: "restano sulla
    # Voce, senza più alcuna Lettura associata") — issue #5.
    insight_senza_lettura: list[InsightEssenziale]


class AggiungiVoceResponse(BaseModel):
    """Risposta di POST /voci. `already_existed` distingue una Voce
    appena creata da una già presente per (utente, libro) — PRD: "se il
    Libro è già in libreria, l'app non lo duplica" — così il chiamante
    decide da solo cosa fare in base allo `stato` ricevuto (in questa
    issue nessun frontend consuma questa distinzione, l'endpoint resta
    testabile via Swagger/API in attesa della ricerca di #4)."""

    voce: VoceResponse
    already_existed: bool


class CambiaStatoRequest(BaseModel):
    """Corpo di PATCH /voci/{id}/stato. `data` è opzionale: se assente,
    la RPC usa il giorno corrente in Europa centrale (PRD: "il giorno
    corrente come predefinito"). A seconda della transizione rappresenta
    la data di inizio (apertura di una Lettura) o di fine (chiusura)."""

    stato: StatoVoce
    data: date | None = None


class CorreggiPagineRequest(BaseModel):
    pagine_adottate: int | None = Field(default=None, gt=0)


class CorreggiVotoRequest(BaseModel):
    """`null` cancella il voto. Mezze stelle ammesse: il vincolo "1-5, a
    scatti di 0,5" ripete a livello applicativo `chk_voce_di_libreria_voto`
    (AGENTS.md: "validazione lato server sempre, anche se il client valida
    già")."""

    voto: float | None = Field(default=None, ge=1, le=5)

    @field_validator("voto")
    @classmethod
    def _voto_a_mezze_stelle(cls, value: float | None) -> float | None:
        if value is not None and (value * 2) % 1 != 0:
            raise ValueError("Il voto ammette solo mezze stelle (1, 1,5, 2, ... 5).")
        return value


class CorreggiNotaIntenzioneRequest(BaseModel):
    """`null` cancella la nota. Nessun limite di lunghezza: il PRD non ne
    impone uno ("nessun limite imposto, quindi nessun rifiuto previsto",
    casi limite)."""

    nota_intenzione: str | None = None

File: frontend/src/lib/api/voci.ts
/**
 * Fetcher per `/voci` e `/voci/{id}` sul backend FastAPI (libreria
 * personale e ciclo di lettura, issue #2). Stesse convenzioni di
 * `lib/api/me.ts`: unioni discriminate per ogni esito, mai un'eccezione
 * per il flusso di controllo, mapping snake_case -> camelCase a carico
 * di questo modulo.
 */

import type { Recensione, RecensioneBody } from "@/lib/api/recensioni";
import { toRecensione } from "@/lib/api/recensioni";
import type { InsightEssenziale, InsightEssenzialeBody } from "@/lib/api/insight";
import { toInsightEssenziale } from "@/lib/api/insight";
import { intestazioniConLingua } from "@/lib/lingua";
import type { ErroreApi } from "@/lib/api/errore";
import { ERRORE_CONFIGURAZIONE, ERRORE_RETE, erroreDaRisposta, regola } from "@/lib/api/errore";

export type StatoVoce = "da_leggere" | "in_lettura" | "in_pausa" | "abbandonato" | "letto";

export type Autore = {
  id: string;
  nomeCanonico: string;
};

export type StatoCopertina = "in_attesa" | "presente" | "assente" | "fallita";

export type Genere = {
  id: string;
  etichetta: string;
};

export type Libro = {
  id: string;
  titoloCanonico: string;
  annoPrimaPubblicazione: number | null;
  /** Vero quando il valore viene dal modello e non dal catalogo/Wikidata.
   * L'etichetta "dedotto" prevista da design doc §9 è stata costruita e
   * poi tolta dall'interfaccia (emendamento 22 agosto 2026): il campo
   * resta esposto, non più mostrato in scheda. */
  annoDedotto: boolean;
  linguaOriginale: string | null;
  linguaDedotta: boolean;
  /** Fino a tre (PRD), mai correggibili da app: "nessun affordance di
   * modifica" è il messaggio, non solo l'assenza di un comando (design
   * doc §9). */
  generi: Genere[];
  /** URL firmato, non il percorso interno: il bucket delle copertine è
   * privato (PRD regola 6) e un percorso da solo non apre nulla. La firma
   * dura sette giorni ed è stabile tra le richieste, così il browser può
   * davvero metterla in cache. */
  copertinaMiniaturaUrl: string | null;
  copertinaGrandeUrl: string | null;
  /** Colore dominante estratto dalla copertina, per ombra e fondo del
   * volume sullo scaffale (design doc §7). Null finché la copertina non è
   * stata recuperata: in quel caso si ricade sul colore derivato dall'id
   * (`lib/spine-color.ts`). */
  copertinaColoreDominante: string | null;
  /** Variante desaturata di `copertinaColoreDominante` per la stanza
   * scura (design doc §3). Null esattamente quando lo è la prima. */
  copertinaColoreDominanteScuro: string | null;
  /** `in_attesa` mentre il lavoro in secondo piano sta recuperando la
   * copertina: è ciò che permette allo scaffale di sapere che vale la
   * pena ricontrollare, invece di aspettare per sempre un'immagine che
   * forse non arriverà. */
  copertinaStato: StatoCopertina;
  autori: Autore[];
};

export type Voce = {
  id: string;
  /** Proprietario della Voce: distingue "è mia" da "è di un collegato"
   * quando la stessa scheda (/libro/[id]) è raggiunta via un collegamento
   * attivo (issue #3). */
  utenteId: string;
  libroId: string;
  stato: StatoVoce;
  pagineAdottate: number | null;
  voto: number | null;
  notaIntenzione: string | null;
  creatoAt: string;
  aggiornatoAt: string;
  /** Pagina dell'ultimo avanzamento della Lettura aperta. Valorizzata solo
   * da GET /voci (filo di avanzamento sulla fascia "in corso", design doc
   * §7): altrove resta null per costruzione della query, non per assenza
   * di dato. */
  paginaCorrente: number | null;
  /** Conteggi, non contenuto: per "Nella tua libreria" ("una recensione,
   * tre insight"). */
  haRecensione: boolean;
  numeroInsight: number;
};

/** Il Libro come arriva alla SCHEDA (`GET /voci/{id}`): quello dello
 * scaffale più la descrizione dell'opera.
 *
 * La separazione rispecchia quella del backend
 * (`app/schemas/voci.py`, `LibroDaScaffale` / `LibroEssenziale`): lo
 * scaffale non disegna la descrizione in nessun punto, e riceverne una
 * per ogni volume della libreria era il pezzo più pesante della
 * risposta della home. Il tipo lo rende un errore di compilazione
 * invece di un campo vuoto scoperto a schermo. */
export type LibroConDescrizione = Libro & {
  /** Solo nella lingua dell'interfaccia, mai un ripiego su un'altra
   * (design doc §9). */
  descrizione: string | null;
  /** Vero quando il testo è stato riformulato dal modello (espanso se
   * troppo corto, accorciato se troppo lungo) a partire dalla
   * descrizione sorgente (design doc §24, emendamento 21 agosto 2026).
   * L'etichetta di trasparenza in scheda è stata costruita e poi tolta
   * dall'interfaccia (emendamento 22 agosto 2026): il campo resta
   * esposto, non più distinto in scheda dalla citazione letterale della
   * fonte. */
  descrizioneRiformulata: boolean;
};

export type VoceConLibro = Voce & { libro: Libro };

export type Avanzamento = {
  id: string;
  pagina: number;
  data: string;
  generatoAutomaticamente: boolean;
};

export type Lettura = {
  id: string;
  /** Nulla per una lettura registrata a posteriori: chi segna oggi un
   * libro letto nel 2019 non sa quando l'ha cominciato, e l'app non lo
   * deduce (migrazione 20260827160000). */
  dataInizio: string | null;
  dataFine: string | null;
  /** L'annata di conclusione quando il giorno non si conosce. Si esclude
   * con `dataFine`: al più una delle due è valorizzata. Entrambe nulle
   * su una lettura conclusa significa che non se ne conosce la data —
   * il libro è letto, ma non appartiene ad alcun anno. */
  annoFine: number | null;
  esito: "conclusa" | "abbandonata" | null;
  avanzamenti: Avanzamento[];
  /** Raggruppati per Lettura (design doc §10), gating spoiler già
   * applicato — issue #5. */
  insight: InsightEssenziale[];
};

export type VoceDettaglio = Omit<VoceConLibro, "libro"> & {
  /** Con la descrizione, a differenza della voce di scaffale: è questa
   * pagina a disegnarla (`components/libro/scheda.tsx`). */
  libro: LibroConDescrizione;
  letture: Lettura[];
  /** `null` se non scritta, o se privata e chi guarda non è il
   * proprietario (RLS, non un campo booleano applicativo) — issue #5. */
  recensione: Recensione | null;
  /** Insight non legati a nessuna Lettura: scritti prima di iniziare il
   * libro, o orfani di una Lettura poi cancellata — issue #5. */
  insightSenzaLettura: InsightEssenziale[];
};

type AutoreBody = {
  id: string;
  nome_canonico: string;
};

type GenereBody = {
  id: string;
  etichetta: string;
};

type LibroBody = {
  id: string;
  titolo_canonico: string;
  anno_prima_pubblicazione: number | null;
  anno_dedotto: boolean;
  lingua_originale: string | null;
  lingua_dedotta: boolean;
  generi: GenereBody[];
  copertina_miniatura_url: string | null;
  copertina_grande_url: string | null;
  copertina_colore_dominante: string | null;
  copertina_colore_dominante_scuro: string | null;
  copertina_stato: StatoCopertina;
  autori: AutoreBody[];
};

type VoceBody = {
  id: string;
  utente_id: string;
  libro_id: string;
  stato: StatoVoce;
  pagine_adottate: number | null;
  voto: number | null;
  nota_intenzione: string | null;
  creato_at: string;
  aggiornato_at: string;
  pagina_corrente: number | null;
  ha_recensione: boolean;
  numero_insight: number;
};

type LibroConDescrizioneBody = LibroBody & {
  descrizione: string | null;
  descrizione_riformulata: boolean;
};

export type VoceConLibroBody = VoceBody & { libro: LibroBody };

type AvanzamentoBody = {
  id: string;
  pagina: number;
  data: string;
  generato_automaticamente: boolean;
};

type LetturaBody = {
  id: string;
  data_inizio: string | null;
  data_fine: string | null;
  anno_fine: number | null;
  esito: "conclusa" | "abbandonata" | null;
  avanzamenti: AvanzamentoBody[];
  insight: InsightEssenzialeBody[];
};

type VoceDettaglioBody = VoceBody & {
  libro: LibroConDescrizioneBody;
  letture: LetturaBody[];
  recensione: RecensioneBody | null;
  insight_senza_lettura: InsightEssenzialeBody[];
};

function toLibro(body: LibroBody): Libro {
  return {
    id: body.id,
    titoloCanonico: body.titolo_canonico,
    annoPrimaPubblicazione: body.anno_prima_pubblicazione,
    annoDedotto: body.anno_dedotto,
    linguaOriginale: body.lingua_originale,
    linguaDedotta: body.lingua_dedotta,
    generi: body.generi.map((genere) => ({ id: genere.id, etichetta: genere.etichetta })),
    copertinaMiniaturaUrl: body.copertina_miniatura_url,
    copertinaGrandeUrl: body.copertina_grande_url,
    copertinaColoreDominante: body.copertina_colore_dominante,
    copertinaColoreDominanteScuro: body.copertina_colore_dominante_scuro,
    copertinaStato: body.copertina_stato,
    autori: body.autori.map((autore) => ({ id: autore.id, nomeCanonico: autore.nome_canonico })),
  };
}

function toLibroConDescrizione(body: LibroConDescrizioneBody): LibroConDescrizione {
  return {
    ...toLibro(body),
    descrizione: body.descrizione,
    descrizioneRiformulata: body.descrizione_riformulata,
  };
}

function toVoce(body: VoceBody): Voce {
  return {
    id: body.id,
    utenteId: body.utente_id,
    libroId: body.libro_id,
    stato: body.stato,
    pagineAdottate: body.pagine_adottate,
    voto: body.voto,
    notaIntenzione: body.nota_intenzione,
    creatoAt: body.creato_at,
    aggiornatoAt: body.aggiornato_at,
    paginaCorrente: body.pagina_corrente,
    haRecensione: body.ha_recensione,
    numeroInsight: body.numero_insight,
  };
}

export function toVoceConLibro(body: VoceConLibroBody): VoceConLibro {
  return { ...toVoce(body), libro: toLibro(body.libro) };
}

function toAvanzamento(body: AvanzamentoBody): Avanzamento {
  return {
    id: body.id,
    pagina: body.pagina,
    data: body.data,
    generatoAutomaticamente: body.generato_automaticamente,
  };
}

function toLettura(body: LetturaBody): Lettura {
  return {
    id: body.id,
    dataInizio: body.data_inizio,
    dataFine: body.data_fine,
    annoFine: body.anno_fine,
    esito: body.esito,
    avanzamenti: body.avanzamenti.map(toAvanzamento),
    insight: body.insight.map(toInsightEssenziale),
  };
}

function toVoceDettaglio(body: VoceDettaglioBody): VoceDettaglio {
  return {
    ...toVoce(body),
    libro: toLibroConDescrizione(body.libro),
    letture: body.letture.map(toLettura),
    recensione: body.recensione ? toRecensione(body.recensione) : null,
    insightSenzaLettura: body.insight_senza_lettura.map(toInsightEssenziale),
  };
}

type ErrorBody = { detail?: string | { error_code?: string; message?: string } };

function baseUrlOrError(): { baseUrl: string } | { status: "error"; errore: ErroreApi } {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    return { status: "error", errore: ERRORE_CONFIGURAZIONE };
  }
  return { baseUrl };
}

export type VociResult =
  | { status: "ok"; data: VoceConLibro[] }
  | { status: "error"; errore: ErroreApi };

/**
 * GET /voci: lo scaffale, la propria libreria.
 *
 * `acceptLanguage`, opzionale (issue #34): la scelta fra le varianti di
 * titolo/descrizione/etichetta di genere nella lingua dell'interfaccia
 * avviene lato backend (`app/core/lingua.py`) sulla stessa intestazione
 * `Accept-Language`. Dal browser il fetch la manda già da solo — nulla da
 * passare qui; da un Server Component (fetch iniziale della pagina) va
 * inoltrata esplicitamente, perché il fetch del server Next.js non eredita
 * gli header della richiesta in arrivo. Vedi `app/(protected)/page.tsx`.
 */
export async function getVoci(accessToken: string, acceptLanguage?: string): Promise<VociResult> {
  const config = baseUrlOrError();
  if ("status" in config) return config;

  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}/voci`, {
      headers: intestazioniConLingua(accessToken, acceptLanguage),
      // Dato per-utente, mai cacheato tra richieste o utenti diversi.
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }

  const body = (await response.json()) as VoceConLibroBody[];
  return { status: "ok", data: body.map(toVoceConLibro) };
}

export type AggiungiVoceResult =
  | { status: "ok"; data: Voce; alreadyExisted: boolean }
  | { status: "not_found" }
  | { status: "error"; errore: ErroreApi };

/** POST /voci: aggiunge un Libro già presente nel catalogo alla propria
 * libreria. Usata dalla fascia "Nella tua libreria" sul libro di un
 * collegato (design doc §15) — non da una ricerca, che è l'issue #4. */
export async function aggiungiVoce(accessToken: string, libroId: string): Promise<AggiungiVoceResult> {
  const config = baseUrlOrError();
  if ("status" in config) return config;

  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}/voci`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ libro_id: libroId }),
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 404) {
    return { status: "not_found" };
  }
  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }

  const body = (await response.json()) as { voce: VoceBody; already_existed: boolean };
  return { status: "ok", data: toVoce(body.voce), alreadyExisted: body.already_existed };
}

export type VoceDettaglioResult =
  | { status: "ok"; data: VoceDettaglio }
  | { status: "not_found" }
  | { status: "error"; errore: ErroreApi };

/** GET /voci/{id}: la scheda del libro, con Libro e storico delle Letture.
 * `acceptLanguage`, opzionale: vedi il docstring di `getVoci` sopra. */
export async function getVoceDettaglio(
  accessToken: string,
  voceId: string,
  acceptLanguage?: string,
): Promise<VoceDettaglioResult> {
  const config = baseUrlOrError();
  if ("status" in config) return config;

  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}/voci/${voceId}`, {
      headers: intestazioniConLingua(accessToken, acceptLanguage),
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 404) {
    return { status: "not_found" };
  }
  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }

  const body = (await response.json()) as VoceDettaglioBody;
  return { status: "ok", data: toVoceDettaglio(body) };
}

export type ScritturaVoceResult =
  | { status: "ok"; data: Voce }
  | { status: "not_found" }
  // "conflitto" copre sia la transizione non ammessa sia il tetto delle
  // pagine adottate: `errorCode` distingue il caso esatto (vedi
  // docs/adr/0015), il chiamante decide come mostrarlo.
  | { status: "conflitto"; errore: ErroreApi }
  | { status: "error"; errore: ErroreApi };

async function patchVoce(
  accessToken: string,
  path: string,
  body: Record<string, unknown>,
): Promise<ScritturaVoceResult> {
  const config = baseUrlOrError();
  if ("status" in config) return config;

  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}${path}`, {
      method: "PATCH",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 404) {
    return { status: "not_found" };
  }
  if (response.status === 409) {
    const errorBody = (await response.json()) as ErrorBody;
    const detail = typeof errorBody.detail === "object" ? errorBody.detail : undefined;
    return { status: "conflitto", errore: regola(detail?.error_code) };
  }
  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }

  return { status: "ok", data: toVoce((await response.json()) as VoceBody) };
}

/** Quanto si sa della data in cui il libro è stato finito. Serve solo a
 * «l'ho già letto», la lettura registrata a posteriori: chi segna oggi un
 * libro letto anni fa sa il giorno, o solo l'anno, o niente, e l'app non
 * riempie i buchi al posto suo. `giorno` è il comportamento di sempre per
 * ogni altra transizione. */
export type PrecisioneChiusura = "giorno" | "anno" | "ignota";

/** PATCH /voci/{id}/stato: l'unica transizione ammessa dallo stato
 * corrente (design doc §9: "l'interfaccia non offre mai una transizione
 * vietata"), ma il 409 resta gestito qui per difesa in profondità. */
export function cambiaStato(
  accessToken: string,
  voceId: string,
  stato: StatoVoce,
  data?: string,
  precisione: PrecisioneChiusura = "giorno",
  annoFine?: number,
): Promise<ScritturaVoceResult> {
  return patchVoce(accessToken, `/voci/${voceId}/stato`, {
    stato,
    ...(data ? { data } : {}),
    // Si manda solo quando non è il default: un corpo che non la nomina
    // si comporta come prima, ed è quello che fanno tutte le altre
    // transizioni.
    ...(precisione === "giorno" ? {} : { precisione }),
    ...(annoFine !== undefined ? { anno_fine: annoFine } : {}),
  });
}

/** PATCH /voci/{id}/pagine-adottate: correzione del totale (design doc
 * §12, "Correggi il totale"). `pagineAdottate: null` rimuove il totale. */
export function correggiPagine(
  accessToken: string,
  voceId: string,
  pagineAdottate: number | null,
): Promise<ScritturaVoceResult> {
  return patchVoce(accessToken, `/voci/${voceId}/pagine-adottate`, {
    pagine_adottate: pagineAdottate,
  });
}

/** PATCH /voci/{id}/voto: `voto: null` cancella il voto. */
export function correggiVoto(
  accessToken: string,
  voceId: string,
  voto: number | null,
): Promise<ScritturaVoceResult> {
  return patchVoce(accessToken, `/voci/${voceId}/voto`, { voto });
}

/** PATCH /voci/{id}/nota-intenzione: `notaIntenzione: null` la cancella. */
export function correggiNotaIntenzione(
  accessToken: string,
  voceId: string,
  notaIntenzione: string | null,
): Promise<ScritturaVoceResult> {
  return patchVoce(accessToken, `/voci/${voceId}/nota-intenzione`, {
    nota_intenzione: notaIntenzione,
  });
}

export type CancellaVoceResult =
  | { status: "ok" }
  | { status: "not_found" }
  | { status: "error"; errore: ErroreApi };

/** DELETE /voci/{id}: cancella l'intera Voce — letture, avanzamenti,
 * voto, recensione, insight, nota di intenzione, preview personalizzata e
 * indici semantici derivati (issue #33). Stesso schema di risposta di
 * `cancellaLettura` (lib/api/letture.ts). */
export async function cancellaVoce(
  accessToken: string,
  voceId: string,
): Promise<CancellaVoceResult> {
  const config = baseUrlOrError();
  if ("status" in config) return config;

  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}/voci/${voceId}`, {
      method: "DELETE",
      headers: { Authorization: `Bearer ${accessToken}` },
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 404) {
    return { status: "not_found" };
  }
  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }
  return { status: "ok" };
}
